lib/classes/ConfigSchemaHandler/errorMessageBuilder.js

```diff
@@ -0,0 +1,171 @@
+  if (!errors.length) throw new Error('Validation errors array cannot be emptry');
+
+  return errors.map(error => {
+    if (error.friendlyMessage) {
+      return error.friendlyMessage;
+    }
+    return buildDefaultErrorMessage(error);
+  });
+};
+
+const buildDefaultErrorMessage = ajvError => {
+  switch (ajvError.keyword) {
+    case 'additionalProperties': {
+      let additionalProperty;
+      if (ajvError.params && ajvError.params.additionalProperty) {
+        additionalProperty = ajvError.params.additionalProperty;
+      }
+      return `${ajvError.dataPath.slice(1)} ${ajvError.message} (received '${additionalProperty}')`;
+    }
+    default:
+      return `${ajvError.dataPath.slice(1)} ${ajvError.message}`;
+  }
+};
+
+const addUserFriendlyMessage = (errors, userConfig) => {
+  return errors.map(error => {
+    let friendlyMessage;
+
+    // regex matches '.functions['xxx'].events[x].xxx'
+    const isEventParam = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]\.[\w]+$/.test(error.dataPath);
+    if (error.params && error.params.additionalProperty && isEventParam) {
+      const eventName = error.dataPath.split('.')[3];
+      friendlyMessage = `Unsupported parameter '${error.params.additionalProperty}' for '${eventName}' event`;
+    }
+
+    // regex matches '.functions['xxx'].events[x]'
+    const isFunctionEvent = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]$/.test(error.dataPath);
+    if (error.params && error.params.additionalProperty && isFunctionEvent) {
+      friendlyMessage = `Unsupported function event '${
+        error.params.additionalProperty
+      }' at ${error.dataPath.slice(1)}`;
+    }
+
+    const isEventAnyOfErrorObject =
+      error.keyword === 'anyOf' &&
+      error.schemaPath === ANY_OF_EVENT_SCHEMA_PATH &&
+      error.params.suberrors &&
+      error.params.suberrors.length &&
+      error.params.suberrors[0].params &&
+      error.params.suberrors[0].params.additionalProperty;
+
+    if (isEventAnyOfErrorObject) {
+      friendlyMessage = `Unsupported function event '${
+        error.params.suberrors[0].params.additionalProperty
+      }' at ${error.dataPath.slice(1)}`;
+    }
+
+    const configUnderValidation = _.get(userConfig, error.dataPath.slice(1));
+
+    if (isEventAnyOfErrorObject && Object.keys(configUnderValidation).length > 1) {
+      const contextStr = `${Object.keys(configUnderValidation).length} (${Object.keys(
+        configUnderValidation
+      ).join(', ')})`;
+      friendlyMessage = [
+        'Event should contain only one root property,',
+        `but got ${contextStr} at ${error.dataPath.slice(1)}`,
+      ].join(' ');
+    }
+
+    if (error.dataPath === '.package' && error.params && error.params.additionalProperty) {
+      friendlyMessage = [
+        `Unsupported package property '${error.params.additionalProperty}'.`,
+        'Avaiable properties: individually, path, artifact, exclude, include, excludeDevDependencies.',
+      ].join(' ');
```

Comment:
I think we should not address `package` property specifically like that.

I'd tackle this problem more generically. Without above patch error message is `package should NOT have additional properties (received 'unregonizedProperty')`. I would replace all messages of such format to: `Unrecognized property 'x' on 'y'`

Reply:
Ah, I love using general cases. Updated to use them instead of patched ones